### ISSUE TITLE
Tracking Rust rewrite

### ISSUE THREAD

**Issue by mindstorm38** [mindstorm38/portablemc#196]
I'm currently quite bored with Python packaging and environment in general. This is just an idea and I know the work is really huge (already tried PoC in the past), so this is a really long-term discussion. Here is a little summary of pros/cons, and the prerequisites we want. **Feel free to give feedback.**

## Prerequisites
- Keep this repository.
- ~Exactly the same CLI, no breaking changes.~
- Support Windows x86+x86_64+Aarch64, macOS x86_64+Aarch64, Linux/BSD x86+x86_64+Aarch64+Armv6(hf).
- Pre-built binaries/libraries for Windows x86+x86_64+Aarch64, macOS x86_64+Aarch64, Linux x86+x86_64+Aarch64+Armv6(hf)

## Pros
- No dependency hell in Rust (so better CLI, this whole idea originates from me trying to switch to click).
- Transparent change for users.
- Possibility of Python binding, you can keep updating CLI/API with `pip`.
- Lower-level native API (with .so, and for Rust dev, obviously, bringing PMC to a wider audience).
- Faster (this was not a big deal with PMC, but it's good to take).
- Standalone binary (potentially dynamically linked to its sibling .so).
- Better inclusion in the UNIX standard directory structure, potential man-page (even if I prefer exhaustive help messages, support for GNU language?).
- Available to many Linux distros (was already the case with pip, anyway...)

## Cons
- Less portability, and I know that it has been really important in PMC adoption. In my opinion, this will not be that hard to have a great portability, we already have knowledge of important OS/Arch to support, `pip` and other package managers can help us to make this transparent to users.
- Rewriting is really likely to bring old bugs back, however we already have a good test suite for this.
- Takes long time, but we already have the knowledge.

**Comment by Ristovski** [comment]
I really believe this idea is worth pursuing.

While your argument is that a Rust rewrite would be _less_ portable than current Python implementation (since you can run Python almost anywhere), I don't necessarily think this will be a problem.

If anything, a proper Rust rewrite should be _more_ portable in the sense you could provide fully static binaries with no external dependencies - this would essentially eliminate what I think is the only limitation of PMC as it stands now - the need for a working Python installation. With a fully static binary, you would not need to depend on _anything_ at runtime.

The external Python dependency is of course not a problem on Linux (given most distros have python preinstalled etc), but I assume it might be a real problem for the average Windows user.

The fact that PMC is implemented with the UNIX philosophy of "do one thing and it do it well" (which is why its so good, tbh), it would make sense for the 'core' of the launcher to be available as a Rust crate/shared lib, and the CLI interface as a thin wrapper of that. This would also allow users to write their own GUI wrappers etc.

tl;dr: The lack of dependency hell + ability to do static binary builds alone make this effort worthwhile, imo.

**Comment by mindstorm38** [comment]
Thank you for this valuable opinion 👍 Your point on portability is right, and I never thought of Windows, but this is really true for non-python-dev users. And yes, the more I work with UNIX systems (Linux), the more I want the launcher to precisely follow UNIX philosophy, and since it is now on AUR, this is more relevant than ever.

**Comment by HappyGoFishing** [comment]
If this rewrite happens, it would be great if CLI args could be entered in any order. I haven't any experience with rust, but I know that there are great CLI arg parsing libraries on crates.io.

**Comment by mindstorm38** [comment]
Yes that would definitely be the case, for now I'll be going on with the well-known "clap" crate. It is quite heavy but also really powerful. Arguments will be accepted in any order within the sub-command they are defined in, exactly like the current behavior with the python implementation. I've seen some projects using portablemc interfacing with portablemc using the command line, so I really don't want it to break, or at least be backward-compatible (for example I really want a `-V` (--version) argument on the command line, instead of `show about`, which will remain but provide more info maybe).

**Comment by HappyGoFishing** [comment]
I'm only a hobbyist programmer with no real experience, so i don't want assume how easy it is to implement anything, but it would be absolutely amazing if there was a way to download and manage modrinth modpacks via the CLI (essentially how PrismLauncher/MultiMC does).

Prismlauncher does have CLI options for doing this, but it still depends the GUI dependencies for QT6.
I was considering researching how to implement something similar to a "headless modrinth launcher", and it was actually through researching that I found this repo!

Anyway, thanks for replying, good luck!

**Comment by mindstorm38** [comment]
I don't know if this is exactly what you are searching for, but there is a project called [ferium](https://github.com/gorilla-devs/ferium) you can be interested to check. This was already discussed for portablemc, and this would be a bit out of scope, but because I found ferium I abandoned the mod manager idea for PMC.

**Comment by HappyGoFishing** [comment]
I already found Ferium yesterday, it was quite good, but a TUI PrismLauncher type experience would be ideal.
Perhaps if I learn rust more seriously I'll attempt to create one.

**Comment by mindstorm38** [comment]
I'm just thinking again of something like `pmc start curseforge:<pack id>` and here we go the game directly starts a modpack. No joke this would be so cool to avoid the horrible curseforge launcher! This come with many problems because it's better to have a specific work-dir for each modpack, so mods/configs don't override. And I don't know how to handle this.

**Comment by HappyGoFishing** [comment]
you can always use prismlauncher instead of using curseforge directly you know?

**Comment by mindstorm38** [comment]
Oo interesting! You know, I've been using PMC for everything for 3 years now so I'm not really aware of the GUI launcher market

**Comment by HappyGoFishing** [comment]
This is what i was alluding to in my earlier comment, a prismlauncher type TUI would be awesome!

**Comment by mindstorm38** [comment]
Side note, thanks for 300 stars this is insane!

**Comment by MulverineX** [comment]
Perhaps this could be a dependency? https://github.com/modrinth/theseus/tree/master/theseus/src/launcher

**Comment by MulverineX** [comment]
It would be a huge boost to the whole community if you instead contributed that work to theusus. It would probably be easier for you too, because you'd have a starting point (rust wise)

**Comment by mindstorm38** [comment]
It would be a huge boost, I agree, but in this specific case I'm afraid that the current architecture I'm going on with would clash with their architecture. For example, they are parsing the version metadata (called "version info" in their code) ahead of time, this isn't compatible with many backward fixes PMC is doing, also they are massively using async, and I'm not a huge fan of async for this case (except for downloads!). These are just examples, but I think you can understand why I'm afraid of trying to convince people to tweak Theseus' code to fit PMC CLI needs. *However*, I would be really happy to contribute to Theseus, in parallel of PMC if possible, to share my knowledge (PMC code itself can be used as a documentation by the way).

**Comment by mindstorm38** [comment]
Some progress update...
```
[  OK  ] Loaded version 1.21 
[  OK  ] Loaded client
[  OK  ] Loaded and verified 70+0 libraries
[  OK  ] Loaded logger client-1.12.xml
[  OK  ] Downloading... 3 MB/s 100.0% (1/1)
[  OK  ] Loaded and verified 3887 assets 17
[  OK  ] Downloading... 36 MB/s 100.0% (3958/3958)
```

This is actually insane, this download is running in a single thread with an async downloader implementation, I'm pretty sure that there is room for improvement, but this is already insanely fast. This is downloading 866 MB in 24 seconds, for full 1.21, this is not counting the JVM because it's not yet ready, and the 24 MB of the client jar file.

**Comment by mindstorm38** [comment]
> It would be a huge boost, I agree, but in this specific case I'm afraid that the current architecture I'm going on with would clash with their architecture. For example, they are parsing the version metadata (called "version info" in their code) ahead of time, this isn't compatible with many backward fixes PMC is doing, also they are massively using async, and I'm not a huge fan of async for this case (except for downloads!). These are just examples, but I think you can understand why I'm afraid of trying to convince people to tweak Theseus' code to fit PMC CLI needs. _However_, I would be really happy to contribute to Theseus, in parallel of PMC if possible, to share my knowledge (PMC code itself can be used as a documentation by the way).

And regarding this, I went back on some of my positions, mostly on ahead of time parsing, because in the end it's so much more practical.

**Comment by mindstorm38** [comment]
I introduced a standardized cache (see below) to avoid polluting the minecraft directory, for now it will be used for versions and JVM manifests. Basically, this will be used for manifest files that the game don't require in its standard directories.

```rust
let mut file = dirs::cache_dir().unwrap_or(env::temp_dir());
file.push("portablemc-cache");
file.push(url_digest);
```

Note that the URL digest is the SHA-1 of the URL.

**Comment by mindstorm38** [comment]
Standard installation is now complete, now I'll be working on how to launch the game! I've added a much more complex and flexible JVM finding policy, that supports static path (with optional version checking), "System" installs finding (reading PATH, and also the standard Linux `/usr/lib/jvm`) and "Mojang" JVM, System and Mojang can be configured to run before one another, the default is currently (may change in the future) to search a JVM on the system and only then to install Mojang provided one, if possible.

```
[  OK  ] Loaded version 1.21  
[  OK  ] Loaded client
[  OK  ] Loaded and verified 70+0 libraries
[  OK  ] Loaded logger client-1.12.xml
[  OK  ] Loaded and verified 3899 assets 17 
[  OK  ] Loaded JVM version Some("21.0.3") at "C:\\Users\\theor\\AppData\\Roaming\\.minecraft_test\\jvm\\java-runtime-delta\\bin\\javaw.exe" 
```

**Comment by jcurtis06** [comment]
I would **LOVE** to see this converted to Rust. I'm currently using PortableMC to run a Minecraft launcher I'm making with Tauri (using Rust to run the PMC Python script). It'd greatly simplify things if I could directly interact with PMC in Rust. Can't wait to see where this goes!

**Comment by mindstorm38** [comment]
Thank you for the support :) this rewrite is on the right track.

**Comment by sedyh** [comment]
How about using Go for that?
It seems to me that its far more convenient options for writing single-binary cli easy and fast.

**Comment by HappyGoFishing** [comment]
> How about using Go for that? It seems to me that its far more convenient option for writing single-binary cli easy and fast.

both go and rust produce statically linked executables, but since rust is more hipsterish, perhaps it will attract more potential contributes.

**Comment by mindstorm38** [comment]
Honestly I don't know Go and I've a pretty good experience in Rust CLI so it's why I'm going with that language, not really for the potential contribute, your point is right tho.

**Comment by mindstorm38** [comment]
First time launching the game successfully! I need to do a lot of test with all Mojang versions, and then work on reproducing the CLI.

![image](https://github.com/user-attachments/assets/fa89b27a-154b-45d7-a183-06ff7f4cb506)

**Comment by mindstorm38** [comment]
My PC may burn at any time (shared bin directory seems to work just really fine)...
![image](https://github.com/user-attachments/assets/f3226a82-6b75-4019-bbae-bf067d590f4b)

**Comment by mindstorm38** [comment]
![image](https://github.com/user-attachments/assets/4c149dbc-ff80-40a5-ae19-5af31efa2ad5)

**Comment by mindstorm38** [comment]
I just discovered an issue that probably affects the current Python implementation. Older versions that have a map_to_resources are acting weirdly (b1.7.3 in my case) because hey are download resources on their own and so alter the resources on each launch, I wondered if this would be the launchwrapper but no, so it should be the version itself? So I don't know what to do with the launcher, do I still have to download and map resources like I've always done, or do I need to let the game downlaod them itself?

**Comment by theRookieCoder** [comment]
Hello, I just wanted to pop in and say that a CLI Minecraft launcher is a dream come true for me! It's awesome that this is being rewritten in Rust, I'll try to contribute to this rewrite if I can.

**Comment by mindstorm38** [comment]
Beautiful!
![image](https://github.com/user-attachments/assets/6a98098e-f4aa-40ab-a92d-dbf222eac56a)

**Comment by Ristovski** [comment]
Do you plan on exposing C bindings with something like https://github.com/mozilla/cbindgen? (maybe there are other better solutions, this is just the I know about in Rust)

This would allow using the PMC lib from any language that supports FFI.

**Comment by mindstorm38** [comment]
Yes, it's on the todo list! I've never done this, so I'm very curious how it works.

**Comment by mindstorm38** [comment]
> 1.2MB executable, very nice!
By the way, have you compiled it to get this number? If so, that's really cool! On the latest commit, in release mode I get 8.4 MB, which is still very good (CLI is now taking some space I guess), but indeed I'll investigate further how to reduce it before releasing.

**Comment by Ristovski** [comment]
> > 1.2MB executable, very nice!
> 
> By the way, have you compiled it to get this number? If so, that's really cool! On the latest commit, in release mode I get 8.4 MB, which is still very good (CLI is now taking some space I guess), but indeed I'll investigate further how to reduce it before releasing.

Indeed! With the latest commit on release I now too get ~8.1MB (6.4MB stripped). With LTO enabled, this can be brought down to ~5.9MB (5.2MB stripped). Sadly rustc seems to emit a DWARF format that https://github.com/google/bloaty does not understand, which is a great tool for checking which compile unit takes up how much space in the binary (`bloaty -d compileunits ./prog`).

**Comment by Ristovski** [comment]
Taking some overkill suggestions from https://github.com/johnthagen/min-sized-rust:
```
[profile.release]
lto = "fat"
strip = true
codegen-units = 1
panic = "abort"
```
brings release size down to 4.5MB! But IMO not worth the hassle :D

**Comment by mindstorm38** [comment]
For final release I think that codegen-units/strip/lto are quite free optimizations, panic is to discuss ^^

**Comment by mindstorm38** [comment]
Also, we'll need to think about the license to use, because I'm not comfortable using GPL anymore in the Rust ecosystem, I prefer Apache 2/MIT for that, and it should not be a problem (on Rust code only), Python part will remain GPL as long as it exists. But I didn't think of that before, but it's implied that the documentation and basically the whole repo, not only the Python code, is under GPL, so I'll still need to contact anybody that has written documentation. Or maybe I drop everything because documentation will be reworked anyway...

I really want to respect every contributor opinion and properly change the license...

**Comment by mindstorm38** [comment]
Improving search command... If you have any idea on what to add here, don't hesitate, I also plan on improving the syntax of the query string!
![{1A466E1A-C204-4024-A2CD-B64EE519951B}](https://github.com/user-attachments/assets/e4a3d253-ac79-448f-8be0-adfaf00485ed)

**Comment by Ristovski** [comment]
Just something to think about: if the same query returns required/"recommended" java/lwjgl versions, then maybe that might be helpful to include in some cases?

Afaik ARM support requires a certain LWJGL version, so knowing which version you could run would be nice etc.

But then again, might not make sense as the returned versions are likely just "recommended" versions, and small deviations are allowed in some cases - like the case where you can use a slightly newer lwjgl native lib etc. But this might not apply for all versions.

**Comment by mindstorm38** [comment]
It's an excellent idea tbh! However, I don't know how to efficiently implement it, because it would require to download all metadata files, this is possible indeed, but will it be performant? With the new cache system, it should be slow only on the first request.

**Comment by mindstorm38** [comment]
I like this idea, and I had it for a long time (not for versions, but for other game resources, or a command to prune unused libraries...). In the implementation I'm imagining, it would still need to download the version metadata, at least, and the assets index (only if we want to give details on the number of assets and the kind of assets mapping there is).

**Comment by mindstorm38** [comment]
Also, I'll be supporting environment variable for some common arguments (game directories, maybe start --login), I think it will be a huge improvement!

**Comment by mindstorm38** [comment]
Fabric-like versions are now supported! Also various fixes... Remaining milestones are: forge/neoforge, and then auth database.

**Comment by mindstorm38** [comment]
![image](https://github.com/user-attachments/assets/0161bab4-b37a-4a3f-aa75-162a70bfdb61)
I think we just hit a new milestone... (so much test to do omg)...

**Comment by mindstorm38** [comment]
I need your opinion on the start and search commands: currently the start command has a specific syntax for the version name, you can use a colon-separater syntax, such as `forge:` or even `mojang:1.21` (in fact its equivalent to just writing `1.21`). I'm wondering if this would not be better to have specific arguments for that purpose instead of reinventing a parsing system, so basically, instead of writing `portablemc start forge:1.21` we would need to write `portablemc start 1.21 --forge [<loader-version>]`. I don't really like this idea, but because I had it, I'm interested if some of you might have interest in this change (for scripting it's maybe simpler? idk).

**Comment by MulverineX** [comment]
imo either support both, or stick with the shorter syntax

**Comment by mindstorm38** [comment]
> imo either support both, or stick with the shorter syntax

You mean, for both start and search command? Because I think that I'll only do this with search command, if I decide to do so.

**Comment by theRookieCoder** [comment]
I'm not really sure about the search command, but as a terminal user I prefer the shorter colon syntax and I would like that to stay. The flags would be better for scripting though. I see that you're using clap, I see how it would be very difficult to integrate the two methods nicely in code.

**Comment by mindstorm38** [comment]
> After some more thought, I think flags are fine for search. There would be many flags to deal with, and knowing the syntax for them as a user might be less intuitive than just checking the flags on the help page. I think they won't be used as much, so I think the colon syntax isn't necessary there.

You summurized better than I could my own opinion ^^ But yes, the start command will remain the same I think!

**Comment by mindstorm38** [comment]
I think that I'm forced to cancel the goal of keeping the CLI the same, it's too complicated to accomodate for all the bad design I made with the Python version.

Also, I'll drop old Mojang's Yggdrasil authentication, and the only authentication available will be Microsoft (until another auth arrives lol, or until we want to support custom auth, idk). Also, I'll no longer associate any email to a session, the `portablemc login` command will simply start a "device code" auth flow, and the session will be cached in the database (without email!), associated to the account's UUID.

**Comment by mindstorm38** [comment]
MSA account support is no fully implemented, older `login` and `logout` commands have been superseded by a more common `auth`, by default this command starts the authentication flow, but it supports various exclusive flags to list accounts, forget and refresh an account. It's no longer possible to "temp login" as before, now, you must log into your account with the `auth` command before starting the game. To start the game with an MSA account, you must now use the `--auth` (`-a`) flag with one of the following `--username` (`-u`) or `--uuid` (`-i`) that will be used to find the associated account in your local database. 

For example, I first log into my "Mindstorm38" account with `portablemc auth`, and then I start the game with it using `portablemc start -a -u Mindstorm38`, and it works. I personally prefer giving the username as it's simpler, but it's possible to give the UUID, which is guaranteed to work even if the username is changed.

**Comment by theoparis** [comment]
I just tried the latest rust-rewrite commit out and it doesn't let me use java 23 with `--jvm` or `--jvm-policy system`: `JVM version 21 not found`.  This seems to be part of the issue https://github.com/mindstorm38/portablemc/blob/rust-rewrite/rust/portablemc/src/standard/mod.rs#L1120, as replacing it with `let major_version = 23` works fine.

**Comment by mindstorm38** [comment]
Really interesting feedback, thank you a lot! Maybe we should be more flexible about Java version compatibility, Java 8 should not be compatible with any version, but maybe starting from 9 we can just allow any version greater? Also, when using `--jvm` the underlying API call is `standard::Installer::set_jvm_policy(JvmPolicy::Static { file: ..., strict_check: true })`, and we can just flip the strict check to false in order to disable that rejecting logic and just issue a warning, maybe it's better?

**Comment by theoparis** [comment]
Yeah, /usr/lib/jvm is the correct location on both Fedora and Ubuntu with Asahi Linux.

**Comment by mindstorm38** [comment]
@theoparis By the way, do you know where are located JVMs on macOS, and how the directories are organized?

**Comment by mindstorm38** [comment]
I've started to reimplement the integration tests, it seems to be working really fine and is much better than before (in that it checks for the right order of operations and for the correct errors), I really like what I've done.

**Comment by mindstorm38** [comment]
I'm trying to define how I'll package and distribute the different crates, but I don't really know how to organize all of this. Essentially, we have the following artifacts:
- `portablemc` or `portablemc.exe`: The executable, standalone because statically linked.
- `libportablemc.so` or `portablemc.dll`: The main shared library for FFI, standalone, requiring libc on linux.
  - This comes with a `portablemc.h`, later we could provide C++ bindings in that header, but not for now.
- Python binding `portablemc`.

**Comment by mindstorm38** [comment]
The (Rust) CLI is now 100% if I'm not missing anything.
The (Rust) API is near 100% complete, impossible to be more precise.
Python binding is WIP.

**Comment by Ristovski** [comment]
This is great news, truly outstanding work!

**Comment by mindstorm38** [comment]
I've made a lot of progress, the 'msa' module is now properly binded on python, I'm still missing the download module and all event handlers!

**Comment by AinRyzen** [comment]
Hi i used you portable Minecraft package in python for quit some time but about 3 months ago I wanted to learn rust so i did so I really want to use this to code like a launcher or mod launcher using some gui library currently looking at gtk 4 and using modrith and forge api just wanted to say hello and show that there are actually people waiting for this ( i am so excited )

**Comment by mindstorm38** [comment]
👀👀👀 

WIP branch: https://github.com/mindstorm38/portablemc/tree/rust-rewrite
CLI on crates.io: https://crates.io/crates/portablemc-cli
Lib on crates.io: https://crates.io/crates/portablemc

Next step is to have a nice GitHub workflow, and then release the C binding, which is already 50% done I think.

Thanks for your patience!

**Comment by theoparis** [comment]
I forgot to mention this a while ago but if I compile and run `portablemc auth` for the first time and go through the auth process it says
```
[ ERRO ] I/O error: No such file or directory (os error 2)
         At /Users/theo/Library/Application Support/minecraft/portablemc_msa.json

```
I worked around it with mkdir + touch and then rerunning portablemc auth but I totally forgot to report this here so it could be fixed.

**Comment by mindstorm38** [comment]
Ohh nice report; I'll fix this for next beta.

**Comment by mindstorm38** [comment]
@theoparis Should be fixed in latest 5.0.0-beta.1, also fixed the README being absent on [crates.io](https://crates.io/crates/portablemc-cli)

**Comment by mindstorm38** [comment]
First release candidate is available!
https://github.com/mindstorm38/portablemc/releases/tag/v5.0.0-rc.1
https://crates.io/crates/portablemc-cli

Feedback is really welcome, if nothing change in the following weeks, I will release the definitive version. No linux package manager support it yet, we are working on AUR support for now.

**Comment by mindstorm38** [comment]
Changes planned for final release:
- [ ] For fabric loaders, the error is not clear when not using the right version, we get internal reqwest HTTP 400 for unknown loaders/games versions, we should have better errors.
- [ ] When launcher the latest forge loader, the order of versions in `maven-metadata.xml` files have changed for Forge (not NeoForge).
- [ ] Remove all directly exposed reqwest API?

**Comment by mindstorm38** [comment]
> ```
> error: invalid value '/usr/lib64/libglfw.so.3.4' for '--exclude-lib <FILTER>': invalid exclude lib pattern, expected <group>:<artifact>:<version>[:<classifier>][@<extension>]
> 
> For more information, try '--help'.
> [rsg245@lap3 sb]$ ./portablemc start ^C
> [rsg245@lap3 sb]$ ./portablemc start --include-bin /usr/lib64/libglfw.so.3.4
> error: unexpected argument '--include-bin' found
> 
>   tip: a similar argument exists: '--include-natives'
> 
> Usage: portablemc start --include-natives <PATH> [VERSION]
> 
> For more information, try '--help'.
> [rsg245@lap3 sb]$ ./portablemc start --include-natives /usr/lib64/libglfw.so.3.4
> ```
> 
> Do not work

Hello, so regarding that, you are almost correct! The detail here is that you are missing the --exclude-lib flag for excluding the natives library that comes with the game, using something like that: `org.lwjgl:lwjgl-glfw:*:natives-linux-arm64` or `-arm32`.

> @mindstorm38, need more docs. How download fabric version?

Using `portablemc start fabric:[<game-version>[:[<loader-version>]]]`.

It's worth mentioning that I actually wrote an extensive documentation in the help message, that you can access by using `--help` instead of `-h`, for example `portablemc start --help`.

**Comment by mindstorm38** [comment]
Ok, that's an issue I'm fixing tomorrow, I've explicitly disallowed stars "*" in the specifiers while working on path safety of libraries.

**Comment by rsg245** [comment]
> > > Ah sorry, you should follow the suggestion, try using --include-natives instead
> > 
> > 
> > [rsg245@lap3 release]$ ./portablemc start --exclude-lib org.lwjgl:lwjgl-glfw:*:natives* --include-natives /usr/lib64/libglfw.so.3.4 fabric:1.21.10
> > error: invalid value 'org.lwjgl:lwjgl-glfw:*:natives*' for '--exclude-lib <FILTER>': invalid exclude lib pattern, expected <group>:<artifact>:<version>[:<classifier>][@<extension>]
> > 
> > For more information, try '--help'.
> 
> This

nice, it works

and new bug

``` shell
 ./portablemc start fabric:1.21.10
[  OK  ] Loaded version fabric-1.21.10-0.18.2
[  OK  ] Fetched version 1.21.10 -- 50.6 kB/s 44 kB (1)
[  OK  ] Loaded version 1.21.10
[  OK  ] Loaded client
[  OK  ] Loaded and verified 78+0 libraries
[  OK  ] Loaded logger client-1.21.2.xml
[  OK  ] Loaded and verified 4397 assets 27 -- 336.9 kB/s 507 kB (1)
[  OK  ] Loaded JVM (21.0.9)
[  ..  ] Downloading -- 8.7 MB/s 539 MB (4477)
[ ERRO ] Failed to download 1 out of 4477 entries...
[ ERRO ] https://maven.fabricmc.net/net/fabricmc/fabric-loader/0.18.2/fabric-loader-0.18.2.jar
         -> /home/rsg245/.minecraft/libraries/net/fabricmc/fabric-loader/0.18.2/fabric-loader-0.18.2.jar
            Reqwest error: error decoding response body
            Unknown error: error reading a body from connection
            I/O error: Connection timed out (os error 110)
./portablemc start fabric:1.21.10
[  OK  ] Loaded version fabric-1.21.10-0.18.2
[  OK  ] Loaded version 1.21.10
[  OK  ] Loaded client
[  OK  ] Loaded and verified 78+0 libraries
[  OK  ] Loaded logger client-1.21.2.xml
[  OK  ] Loaded and verified 4397 assets 27
[  OK  ] Loaded JVM (21.0.9)
[  OK  ] Launched
Error: Could not find or load main class net.fabricmc.loader.impl.launch.knot.KnotClient
Caused by: java.lang.ClassNotFoundException: net.fabricmc.loader.impl.launch.knot.KnotClient
./portablemc start fabric:1.21.10
[  OK  ] Loaded version fabric-1.21.10-0.18.2
[  OK  ] Loaded version 1.21.10
[  OK  ] Loaded client
[  OK  ] Loaded and verified 78+0 libraries
[  OK  ] Loaded logger client-1.21.2.xml
[  OK  ] Loaded and verified 4397 assets 27
[  OK  ] Loaded JVM (21.0.9)
[  OK  ] Launched
Error: Could not find or load main class net.fabricmc.loader.impl.launch.knot.KnotClient
Caused by: java.lang.ClassNotFoundException: net.fabricmc.loader.impl.launch.knot.KnotClient
```

**Comment by mindstorm38** [comment]
It is definitely planned for future v5 versions, I really want this in the launcher!

**Comment by mindstorm38** [comment]
Maybe something we want for the final release is a better readme with examples?

**Comment by rsg245** [comment]
@mindstorm38, Support codberg.org you can easily create a mirror of your repository there, this will reduce delays to the server (German server), add ipv6 support and make it possible to download portablemc when github is unavailable

**Comment by rsg245** [comment]
> We already have access to a AUR repo for which I no longer have news at the moment. I contacted the maintainer a month ago I think

Did the attendant respond? I have sent a request to AUR for access to push updates, but it has not been responding for a month. Could you ask him to add me to contributors? My nickname in AUR: `rsg245`  [Link](https://aur.archlinux.org/account/RSG245/)

**Comment by mindstorm38** [comment]
> Did the attendant respond? I have sent a request to AUR for access to push updates, but it has not been responding for a month. Could you ask him to add me to contributors? My nickname in AUR: rsg245 Link

I've tried at some point to report the package as abandonned and at this time the maintainer contacted me, and so I closed the request. Since the. I've no news.

> @mindstorm38, successful migration to [codeberg](https://codeberg.org/portablemc/portablemc). Check your access to org and repos

It's ok, ho, is it managed? I have read that code berg have disabled creation of mirrors, so I wonder if I'll need to add some GitHub workflow to synchronize the repo there?

> in general, there is an extended functionality compared to github. There are no paid features. This is the largest forgejo instance in Europe.

Ok I see, I like this idea honestly and having a mirror is not bad anyway. Also, I've seen that the mirror have the merge requests clones, I'll disable this because for now I still want to manage everything on GH.

Ah and my question was more about the PKGBUILD, are you able to use two URLs for the artifact or the source code in case GH goes down?

Btw I think we'll need two AUR packages for prebuilt and from source? Is it how we're supposed to organize?

**Comment by rsg245** [comment]
@mindstorm38, i can make portablemc-bin aur package from x86_64-unknown-linux-gnu.zip archive. 

Can you build linux packages with standart names? portablemc-{version}-x86_64.tar.gz (archive tar + gzip2 or bzip2). 

And Sign your files in releases. I need an archive.sig (upload with .tar.gz archive in GH) file with your pgp key fingerprint. You can publish your pub pgp key on keyserver.ubuntu.com

**Comment by mindstorm38** [comment]
By the way I'll be doing all this after holidays :)

**Comment by mindstorm38** [comment]
Version 5.0.0 is out!!

Release: https://github.com/mindstorm38/portablemc/releases/tag/v5.0.0
Crates.io: https://crates.io/crates/portablemc-cli

Now merged on the main branch!

Please open standalone issues now!

**Comment by rsg245** [comment]
@mindstorm38, ASCII format is not used in pacman and the automatic process. all package mirrors for arch and other aur pkbulds use .sig

**Comment by rsg245** [comment]
@mindstorm38, and use zip archive for windows pkgs

**Comment by mindstorm38** [comment]
Arghhh fuck, okay, I'll sign it manually and reupload everything... It was intentional.

And for windows package honestly I don't really care, I know that I should've used ZIP but modern windows is supporting tar.gz natively in the file explorer. Older versions should use 7zip anyway. At least for this 5.0.0 release...

**Comment by mindstorm38** [comment]
I agree.

I've uploaded the .sig just now.

**Comment by mindstorm38** [comment]
@rsg245 I've implemented all those fix to branch `fix-v5-workflow`, I will wait before releasing a patch, as it is not urgent. So ZIP for windows, and using no-ascii signatures. I go sleep now :) 

Ah and maybe we can wait before patching, so we can also test the release automation on your side for the AUR?
